### DISCUSSION TITLE
Enable Link-Time Optimization (LTO)

### DISCUSSION THREAD

**Discussion by zamazan4ik** [ilai-deutel/kibi#305] [Ideas]
Hi!

I noticed that in the `Cargo.toml` file Link-Time Optimization (LTO) for the project is not enabled. I suggest switching it on since it will reduce the binary size (always a good thing to have) and will likely improve the application's performance a bit.

I suggest enabling LTO only for the Release builds so as not to sacrifice the developers' experience while working on the project since LTO consumes an additional amount of time to finish the compilation routine. If you think that a regular Release build should not be affected by such a change as well, then I suggest adding an additional `dist` or `release-lto` profile where additionally to regular `release` optimizations LTO will also be added. Such a change simplifies life for maintainers and others interested in the project persons who want to build the most performant version of the application. Using ThinLTO should also help to reduce the build-time overhead with LTO. If we enable it on the Cargo profile level, users, who install the application with `cargo install`, will get the LTO-optimized version "automatically". E.g., check `cargo-outdated` Release [profile](https://github.com/kbknapp/cargo-outdated/blob/master/Cargo.toml#L48).

Basically, it can be enabled with the following lines:

```
[profile.release]
lto = true
```

I have made quick local tests (Rustc 1.81, Fedora 40, the latest `kibi` version): with `lto = true` the `kibi` binary size dropped from 641 Kib to 576 Kib.

Thank you.

**Comment by ilai-deutel** [comment]
Hi @zamazan4ik,

Sorry for the delay, and thank you for the suggestion! Adding LTO to the release profiles sounds like a great idea, feel free to open a pull request if you'd like.

**Reply by zamazan4ik** [reply]
Sure: https://github.com/ilai-deutel/kibi/pull/346
